### PR TITLE
Fix hashing

### PR DESCRIPTION
**Pull request by woodruffw**
This reverts commit d16e9ec2b4d54c1333137ae34f011e9eb3e0b962.

CC @pkdawson: This is going to revert your work in #62 due to some issues around [Authenticode stuffing](https://vcsjones.dev/authenticode-stuffing-tricks/) that were recently brought to our attention. 

My plan is to re-merge your fix, tweaked to perform the section sorting the way we originally did it.

### CODE REVIEW

**Comment by woodruffw** [comment]
> Marking as draft until I fix the hash calculation for unsigned binaries.

I'm just going to punt on this entirely and update the API to not support hashing unsigned binaries: doing so isn't well defined in the Authenticode spec anyways.

**Comment by pkdawson** [comment]
Testing this branch, I see a regression for #59, where svcli shows the same incorrect calculated checksum for the linked Firefox installer.

I don't understand this issue well enough to give useful input, but for what it's worth, [CryptQueryObject](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptqueryobject) fails to find an embedded signature in `YourPhone.exe` (error code `CRYPT_E_NO_MATCH`). This API does not check the Authenticode hash.

**Comment by woodruffw** [comment]
Thanks for checking that -- I need to add a regression test for that as well. I won't merge until we get the correct hash for PEs that _do_ contain signatures.

File: src/uthenticode.cpp [hunk 1]
```diff
@@ -541,36 +551,53 @@ std::string calculate_checksum(peparse::parsed_pe *pe, checksum_kind kind) {
...
-  auto *md_ctx = EVP_MD_CTX_new();
-  EVP_DigestInit(md_ctx, md);
-  EVP_DigestUpdate(md_ctx, pe_bits.data(), pe_bits.size());
-
-  if (security_dir.VirtualAddress > 0) {
-    /* If a certificate table exists, hash everything before and after it.
-     */
-    EVP_DigestUpdate(md_ctx,
-                     pe->fileBuffer->buf + size_of_headers,
-                     security_dir.VirtualAddress - size_of_headers);
```
Comment (ret2libc) [inline src/uthenticode.cpp:556]: Have you removed this part?

Comment (woodruffw) [reply src/uthenticode.cpp:556]: Sort of -- part of this is covered by the section iteration, but not all of it (which is why we've regressed on some PEs, as noted in https://github.com/trailofbits/uthenticode/pull/84#issuecomment-1660794391).

I need to think more about an appropriate way to do this -- we need to perform the sorted section iteration *and* hash everything after the sections.

**Comment by woodruffw** [comment]
With the latest changes I now get the expected hashes for the testcase in #59:

```
./src/svcli/svcli ~/Downloads/Firefox\ Setup\ 101.0.exe        
This PE is verified!                                                                              
                                                                                                  
/Users/william/Downloads/Firefox Setup 101.0.exe has 1 certificate entries                        
                                                                                                  
Calculated checksums:                                                                             
   MD5: 0678B4315E48B222D3B54477E51C64F0                                                          
  SHA1: 69A54D16B3E176E16EF6AE5BBC6490F9F019A9B6                                                  
SHA256: DEF2F92C51E7DD222F480B9459C1E2E0A35C598453FC85EF74D9439D105AC853                                                                                              
```

I'm going to look into a smaller PE that I can include in the test suite to prevent regressions here.

**Pull request merged**